articles/051_actions.md

```diff
@@ -0,0 +1,95 @@
+
+This document defines how actions are specified, what they look like to ROS users (both node developers and system integrators)
+
+## Action specification
+
+Actions are specified using a form of the ROS Message IDL. The specification contains three sections, each of which is a message specification:
+
+1. Goal
+
+1. Result
+
+1. Feedback
+
+Any of these sections may be empty.
+
+Between the three sections is a line containing three hyphens, `---`.
+
+Action specifications are stored in a file ending in `.action`. There is one action specification per `.action` file.
+
+An example action specification [taken from the actionlib wiki] is shown below.
+
+```
+# Define the goal
+uint32 dishwasher_id  # Specify which dishwasher we want to use
+---
+# Define the result
+uint32 total_dishes_cleaned
+---
+# Define a feedback message
+float32 percent_complete
+uint32 number_dishes_cleaned
+```
+
+## Serving and using actions
+
+Actions are a first-class citizen in the ROS API, alongside topics and services.
+
+Action clients will use an API that provides a proxy object for the action. This will be a templated class, using the action class generated from the action specification as the template parameter. The client shall create an instance of this class, providing the address of the intended action server. Each instance of this class can only be related to one action server. Methods of the class will provide facilities for sending a goal to the action server, receiving a result, and getting feedback.
+
+Action servers will use an API that provides a templated server class, using the action class generated from the action specification as the template parameter. The node implementer will create a function that implements the action's behaviour, create an instance of the templated server class, and bind the implementing function to the server. The implementing function will receive as one of its parameters the received goal message, and as another parameter the action server instance. The implementation shall use the action server instance to provide progress feedback and to report the result and success/failure/error status of the action's execution.
+
+Actions may be used from or served by real-time nodes. Therefore the actions API must be real-time capable.
+
+## Introspection tools
+
+Actions, like topics and services, are introspectable from the command line.
+
+In ROS 1, actions are visible in the output of the `rostopic` tool. 
+
+In ROS 2, actions will not be visible as a set of topics. Nor will they be visible as a set of services [in the case that services be used to implement them]. They will be visible using a separate `ros2 action` command line tool.
```

Comment:
There is the notion of hidden topics and services. Topics/services starting with an underscore `_` are considered "hidden".
Design doc reference: http://design.ros2.org/articles/topic_and_service_names.html#hidden-topic-or-service-names
The commandline tools will not show these topics by default but should provide a flag to display  them.

Reply:
@mikaelarguedas - Thank you for the fast answer. So given that is the case, can Action servers be implemented as a class that internally creates both a hidden service and a hidden topic? Likewise an action client would call that service and subscribe to that topic.

Reply:
Yes I think this is doable :+1:

Reply:
@mikaelarguedas - given that is true, there are at least 2 ways to implement this I can think of:
1) Port actionlib and actionlib_msgs directly, except adding the "hiding" of topic and service names
- This would be the most straight 'port' possible
- Would still need to update the ros2utils to add the 'ros2 actions' functions. This is needed no matter what option is chosen
- don't know what colcon / ament build changes are needed if any
2) Create new ActionServer and ActionClient classes for ROS2, with create_action_server and create_action_client as part of the node interface available in rclpy and rclcpp
 - This would make Actions available through the node interface directly, just like Topics and Services
- They could be implemented internally as 'hidden' topics and services 
- still need to add the 'ros2 actions' functions
- don't know what colcon / ament build changes are needed if any

Any other options? I think Actions could be plumbed down all the way to the RMW layer but I don't know how beneficial that would be, and it might result in unneeded code duplication of what is in Services and Topics today. 

Please share your thoughts on this. My team is going to be blocked by this for Navigation, but if we can figure out the design, we could potentially help with some of the implementation work.

Reply:
I'm in favour of option 2, because I want to see actions be treated as first-class citizens like topics and services in ROS 2, not the ignored illegal alien that they are in ROS 1 tooling.

A simple implementation would be to just use hidden topics and services, and I see this as being the easiest way to go to prototype the API. Since the implementation and the high-level API/tooling design should be relatively independent, this could be a good way to get something useable out. However, I wouldn't want this to lead to stagnation that results in a better implementation that takes advantage of available `rmw`/DDS features not being produced. I'm very keen to hear what the OSRF people had thought about how DDS would be leveraged for actions.